src/filter.rs

```diff
@@ -0,0 +1,300 @@
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error};
+use std::ffi::{CString, CStr};
+use std::path::PathBuf;
+use raw::{git_filter, git_writestream};
+
+
+#[repr(C)]
+struct Filter<T> {
+    filter: git_filter,
+    instance: T,
```

Comment:
Yeah, i'm struggling whether i should use trait object here, maybe there won't be so such different kinds of filters and monomorphization costs are not so high?